### ISSUE TITLE
restoring ckan and datastore databases with docker commands?

### ISSUE THREAD

**Issue by Analect** [ckan/ckan-docker#14]
@clementmouchet 
In order to restore a backup of the ckan and datastore databases from ckan 2.2, is there any special technique required to do this with this docker set-up?

In the past, I have used some the following with psql, but I'm not sure if that is sufficient?  Do I need to manually alter passwords on this ckan-docker set-up to align with the passwords I was using on my previous verison of the database?

docker run --rm --interactive --link db:db --volume $PWD/:/tmp/ ckan/postgresql bash -c 'PGPASSWORD=xxxx exec psql -h "172.17.0.10" -p "5432" -U ckan_default -d datastore_default < /tmp/20141030_datastore_default.bak'

Alternatively, another docker set-up has been able run paster commands from temporary CKAN instances, such as below.  Is that a better approach to getting the old databases upgraded to 2.3?

Obviously this paster is related to rebuilding a solr index, but an equivalent one to upgrade a database would use this 'paster db upgrade' utility. 
http://docs.ckan.org/en/latest/contributing/database-migrations.html

docker run --rm -i -t --link db:db --link solr2.2:solr cygri/ckan:2.2 /sbin/my_init -- /bin/bash -c '$CKAN_HOME/bin/paster --plugin=ckan search-index rebuild -r -c $CKAN_CONFIG/ckan.ini'

Thanks for your insights.

**Comment by clementmouchet** [comment]
I prefer to run commands using docker exec -it so that the whole stack is already up and linked. I suggest you use fig or docker-compose to do you run commands, as it'll take care of the links that you need.

As far as the command is concerned, you should use the hostname rather than the IP, because it'll remain the same. 

I use the following command with cron to backup the database (from the CKAN container)
```
PGUSER=$POSTGRES_ENV_CKAN_USER PGPASSWORD=$POSTGRES_ENV_CKAN_PASS pg_dump -h postgres $POSTGRES_ENV_CKAN_DB > $CKAN_DATA/backups/`date +"%Y_%m_%d-%H%M"`_$POSTGRES_ENV_CKAN_DB.sql

PGUSER=$POSTGRES_ENV_CKAN_USER PGPASSWORD=$POSTGRES_ENV_CKAN_PASS pg_dump $POSTGRES_ENV_DATASTORE_DB > $CKAN_DATA/backups/`date +"%Y_%m_%d-%H%M"`_$POSTGRES_ENV_DATASTORE_DB.sql
```

and here's how to restore it (from the Postgres Container)
```
# Drop CKAN & Datastore databases
setuser postgres psql -c "DROP DATABASE IF EXISTS $CKAN_DB;"
setuser postgres psql -c "DROP DATABASE IF EXISTS $DATASTORE_DB;"
 
# create CKAN & Datastore databases
setuser postgres createdb \
  -O $CKAN_USER $CKAN_DB -T template0 -E utf-8
setuser postgres createdb \
  -O $CKAN_USER $DATASTORE_DB -T template0  -E utf-8
 
setuser postgres psql -d $CKAN_DB < /var/lib/ckan/backups/2015_02_26-2124_dump.sql
setuser postgres psql -d $DATASTORE_DB < /var/lib/ckan/backups/2015_02_26-2124_dump.sql
```

hope this helps. I'll update the doc as soon as possible, but in the meantime pull requests are welcome :)

**Comment by Analect** [comment]
@clementmouchet 
I'm struggling a bit here, so any pointers would be much appreciated.  Your backup / restore approach above assumes the databases are already in a compatible form for the set-up defined in this docker environ. My CKAN installation has been working on a different docker set-up and so has been a bit more complicated to migrate. It's also not clear if I should be somehow upgrading my database on restoring, given the possible structure changes brought about by resource views etc.. 

I was using the cygri/ckan2.2 docker solution, which used various users/owners (ckan, ckan_default, datastore_default) that differ from yours (ckan_user, datastore_user).
I thought I had managed to change all these ahead of generating a pg_dump of each of ckan and datastore databases.

Ref the ckan database, I was able to pg_dump as follows.
docker run --rm --interactive --link db:db --volume $PWD/:/tmp/ ckan/postgresql bash -c 'PGPASSWORD=ckan_pass exec pg_dump -h "172.17.0.2" -p "5432" -U ckan_user ckan -F tar -f /tmp/ckan_20150318.tar'

I then copied this tar to a /tmp/ directory in the new ckandocker_postgres_1 and was able to run:
pg_restore -c -Ft ckan_20150318.tar -d ckan

This appeared to broadly work ... except for a string of errors like these
pg_restore: [archiver (db)] Error from TOC entry 161; 1259 18031 TABLE activity ckan_user
pg_restore: [archiver (db)] could not execute query: ERROR:  table "activity" does not exist
    Command was: DROP TABLE public.activity;
WARNING: errors ignored on restore: 267

While the various datasets were not initially visible, I ran:
paster search-index rebuild -r -c /etc/ckan/default/ckan.ini
back on the ckandocker_ckan_1 container.

This appeared to resolve things on the ckan database, at least.

I then tried restoring the much larger datastore database in a similar approach, but hit problems there with user ckan_default still associated with tables ... and will need to fix this.
pg_restore: [archiver (db)] Error from TOC entry 227; 1259 19554 TABLE fcd80b3f-7de6-42bd-81d3-0c25223dcf03 ckan_default
pg_restore: [archiver (db)] could not execute query: ERROR:  role "ckan_default" does not exist
    Command was: ALTER TABLE public."fcd80b3f-7de6-42bd-81d3-0c25223dcf03" OWNER TO ckan_default;

This somehow made nginx unresponsive, so I decided to exit the 'docker compose up' process and restart ... but now that won't even restart, with error logs, as set out below.

Any idea how I might resolve this ... without having to start from scratch?

I think those (including me!) hoping to leverage this ckan-docker approach are going to need some help with migrating databases from older set-ups (using ckan_default and datastore_default).

Thanks. Colum

ckan_1       | *** Running /etc/my_init.d/70_initdb...
solr_1       | 6967 [qtp883065319-11] INFO  org.apache.solr.core.SolrCore  – [ckan] webapp=/solr path=/select params={fl=*,score&q=*:*&wt=standard&version=2.2&rows=1} hits=0 status=0 QTime=2 
solr_1       | 6974 [qtp883065319-12] INFO  org.apache.solr.core.SolrCore  – [ckan] webapp=/solr path=/admin/file/ params={file=schema.xml} status=0 QTime=0 
solr_1       | 7152 [qtp883065319-13] INFO  org.apache.solr.core.SolrCore  – [ckan] webapp=/solr path=/select params={fl=*,score&q=*:*&wt=standard&version=2.2&rows=1} hits=0 status=0 QTime=0 
solr_1       | 7155 [qtp883065319-14] INFO  org.apache.solr.core.SolrCore  – [ckan] webapp=/solr path=/admin/file/ params={file=schema.xml} status=0 QTime=0 
postgres_1   | ERROR:  no schema has been selected to create in
postgres_1   | STATEMENT:  CREATE OR REPLACE VIEW "_table_metadata" AS 
postgres_1   | 	            SELECT DISTINCT
postgres_1   | 	                substr(md5(dependee.relname || COALESCE(dependent.relname, '')), 0, 17) AS "_id",
postgres_1   | 	                dependee.relname AS name,
postgres_1   | 	                dependee.oid AS oid,
postgres_1   | 	                dependent.relname AS alias_of
postgres_1   | 	                -- dependent.oid AS oid
postgres_1   | 	            FROM
postgres_1   | 	                pg_class AS dependee
postgres_1   | 	                LEFT OUTER JOIN pg_rewrite AS r ON r.ev_class = dependee.oid
postgres_1   | 	                LEFT OUTER JOIN pg_depend AS d ON d.objid = r.oid
postgres_1   | 	                LEFT OUTER JOIN pg_class AS dependent ON d.refobjid = dependent.oid
postgres_1   | 	            WHERE
postgres_1   | 	                (dependee.oid != dependent.oid OR dependent.oid IS NULL) AND
postgres_1   | 	                (dependee.relname IN (SELECT tablename FROM pg_catalog.pg_tables)
postgres_1   | 	                    OR dependee.relname IN (SELECT viewname FROM pg_catalog.pg_views)) AND
postgres_1   | 	                dependee.relnamespace = (SELECT oid FROM pg_namespace WHERE nspname='public')
postgres_1   | 	            ORDER BY dependee.oid DESC;
postgres_1   | 	        
ckan_1       | Traceback (most recent call last):
ckan_1       |   File "/usr/lib/ckan/default/bin/paster", line 9, in <module>
ckan_1       |     load_entry_point('PasteScript==1.7.5', 'console_scripts', 'paster')()
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/paste/script/command.py", line 104, in run
ckan_1       |     invoke(command, command_name, options, args[1:])
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/paste/script/command.py", line 143, in invoke
ckan_1       |     exit_code = runner.run(args)
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/paste/script/command.py", line 238, in run
ckan_1       |     result = self.command()
ckan_1       |   File "/usr/lib/ckan/default/src/ckan/ckan/lib/cli.py", line 208, in command
ckan_1       |     self._load_config(cmd!='upgrade')
ckan_1       |   File "/usr/lib/ckan/default/src/ckan/ckan/lib/cli.py", line 148, in _load_config
ckan_1       |     load_environment(conf.global_conf, conf.local_conf)
ckan_1       |   File "/usr/lib/ckan/default/src/ckan/ckan/config/environment.py", line 232, in load_environment
ckan_1       |     p.load_all(config)
ckan_1       |   File "/usr/lib/ckan/default/src/ckan/ckan/plugins/core.py", line 134, in load_all
ckan_1       |     load(*plugins)
ckan_1       |   File "/usr/lib/ckan/default/src/ckan/ckan/plugins/core.py", line 166, in load
ckan_1       |     plugins_update()
ckan_1       |   File "/usr/lib/ckan/default/src/ckan/ckan/plugins/core.py", line 116, in plugins_update
ckan_1       |     environment.update_config()
ckan_1       |   File "/usr/lib/ckan/default/src/ckan/ckan/config/environment.py", line 361, in update_config
ckan_1       |     plugin.configure(config)
ckan_1       |   File "/usr/lib/ckan/default/src/ckan/ckanext/datastore/plugin.py", line 117, in configure
ckan_1       |     self._create_alias_table()
ckan_1       |   File "/usr/lib/ckan/default/src/ckan/ckanext/datastore/plugin.py", line 237, in _create_alias_table
ckan_1       |     connection.execute(create_alias_table_sql)
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 719, in execute
ckan_1       |     return self._execute_text(object, multiparams, params)
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 868, in _execute_text
ckan_1       |     statement, parameters
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 954, in _execute_context
ckan_1       |     context)
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1116, in _handle_dbapi_exception
ckan_1       |     exc_info
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 189, in raise_from_cause
ckan_1       |     reraise(type(exception), exception, tb=exc_tb)
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 947, in _execute_context
ckan_1       |     context)
ckan_1       |   File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 435, in do_execute
ckan_1       |     cursor.execute(statement, parameters)
ckan_1       | sqlalchemy.exc.ProgrammingError: (ProgrammingError) no schema has been selected to create in
ckan_1       |  'CREATE OR REPLACE VIEW "_table_metadata" AS \n            SELECT DISTINCT\n                substr(md5(dependee.relname || COALESCE(dependent.relname, \'\')), 0, 17) AS "_id",\n                dependee.relname AS name,\n                dependee.oid AS oid,\n                dependent.relname AS alias_of\n                -- dependent.oid AS oid\n            FROM\n                pg_class AS dependee\n                LEFT OUTER JOIN pg_rewrite AS r ON r.ev_class = dependee.oid\n                LEFT OUTER JOIN pg_depend AS d ON d.objid = r.oid\n                LEFT OUTER JOIN pg_class AS dependent ON d.refobjid = dependent.oid\n            WHERE\n                (dependee.oid != dependent.oid OR dependent.oid IS NULL) AND\n                (dependee.relname IN (SELECT tablename FROM pg_catalog.pg_tables)\n                    OR dependee.relname IN (SELECT viewname FROM pg_catalog.pg_views)) AND\n                dependee.relnamespace = (SELECT oid FROM pg_namespace WHERE nspname=\'public\')\n            ORDER BY dependee.oid DESC;\n        ' {}
ckan_1       | *** /etc/my_init.d/70_initdb failed with status 1
ckan_1       | 
ckan_1       | *** Killing all processes...
ckandocker_ckan_1 exited with code 1
Gracefully stopping... (press Ctrl+C again to force)
Stopping ckandocker_nginx_1...
Stopping ckandocker_postgres_1...
Stopping ckandocker_data_1...
Stopping ckandocker_solr_1...
Stopping ckandocker_datapusher_1...

**Comment by Analect** [comment]
@clementmouchet 

Related to the above, I can see from running 'paster db version' on my current set-up that I'm on version 71. Does that require me to manually apply scripts 72 through 76 on my existing database to get it to be ckan 2.3 compliant ahead of dumping and restoring to this ckan-docker set-up, or is this handled in any way by the db initiation scripts built in to this docker solution?  Is 'paster db upgrade' getting run automatically that figures out that my old restored database is only on version 71?
https://github.com/ckan/ckan/tree/master/ckan/migration/versions

I'm not sure whether the migration docs linked below are current any longer ... is the idea to try to get the old *.ini config file (in terms of user names and database names) aligned as much as possible with the ckan.ini file as per this docker set-up, ahead of dumping the old ckan and datastore databases?
http://docs.ckan.org/en/latest/contributing/database-migrations.html

**Comment by Analect** [comment]
My data is stuck on 2.2 (ckan and datastore databases). Have exhausted efforts to backup and restore correctly.  https://github.com/ckan/ckan/commit/0f7e7f1aa564a87daf1c14328ba1348a562f9eba suggests various upgrade requirements ... but it's not obvious how these get implemented in this docker set-up.
An added complication is that user and database names differ in this implementation from the defaults used in the main ckan repo and cygri/ckan:2.2.

**Comment by Analect** [comment]
OK. I haven't fully solved this, but I'll post some steps here in case others are having trouble moving from another docker set-up and an earlier version of CKAN. While I can now restore my data to vers 2.3 under this docker set-up, I'm still having issues in adding new datasets, which I will elaborate on in a separate issue. Meanwhile I'll close this issue with a summary of my approach below.

1. Was moving from a set-up using ver 2.2 and different database names / users (see in comments above). Renamed databases / users and dumped to an *.sql file using psql.
2. Back in this ckan-docker set-up, prior to running up the containers, I ensured that the ckan repo cloned to '_src/ckan' was on branch 'release v.2.2).
3. This ckan-docker makes use of a make-config paster function that wasn't available in ver 2.2 and so I changed the following files manually as per the niryuu/ckan-docker fork here: https://github.com/niryuu/ckan-docker/commit/fdca1ba448af70f5e08a8fc958bf64408c1b19d8
4. I also had to use an older schema.xml (version 2.0 and not 2.3) under the _solr folder for the full suite of containers to start properly.
5. Now that the ckan-docker setup was is in version 2.2 mode, I am able to restore my previous system data, without error.
6.  I then mainly followed these steps: 
http://docs.ckan.org/en/ckan-2.3/maintaining/upgrading/upgrade-source.html 
 ... within the running ckan container, checking out branch 'release v.2.3.1' of the ckan repo at: /usr/lib/ckan/default/src/ckan.  
7. The jetty restart step in that link doesn't apply, from what I can tell. Or perhaps it's necessary to enter the solr container and issue some restart command.
8. The paster step to reindex datasets was absolutely necessary, but for some reason, this isn't persisted for me in this setup, requiring me to re-run that reindexing if I had to re-do 'docker-compose up' again. I'm not sure if the solr data is getting persisted back to the 'data' container in this set-up?

**Comment by clementmouchet** [comment]
Just catching up on that issue.
Glad you solved it eventually, but it seems you have taken the scenic route ;)


An easier approach would be to specify you current db names & credentials via the environment variables, letting ckan-docker create a 2.3 schema & db withthe same ones during the initial boot.
When that's done, you could stop all CKAN services (apache & any supervisord tasks like celery, and or harvesting tasks to free postgres connections)
You could then import you db, effectively overriding the blank new one, and then runnthe usual migration commands.

Restart the servives and you should be good ;)
